test/Interop/Cxx/operators/read-accessor.swift

```diff
@@ -0,0 +1,45 @@
+// RUN: %target-swift-emit-silgen %s -I %S/Inputs -enable-cxx-interop | %FileCheck %s
+
+import MemberInline
+
+// Verify that operators returning by reference generate a read accessor and call it.
+
+public func index(_ arr: inout ReadOnlyIntArray, _ arg: Int32) -> Int32 { arr[arg] }
```

Comment:
Let me think about this test a little bit. I wonder if there's a way to make sure we don't copy the value, and that we yield a specific value. But this mostly looks good.

Reply:
I revamped this test to check that we don't copy the value. Just for illustrative purposes here, we see the old getter and the new read accessor side by side. Importantly, the getter has a `copy_addr` instruction
that is left out in the reader.

**Previous Getter:**
```sil
// NonTrivialArrayByRef.subscript.getter
sil shared [transparent] @$sSo20NonTrivialArrayByRefVySo0aB0Vs5Int32Vcig : $@convention(method) (Int32, @inout NonTrivialArrayByRef) -> @out NonTrivial {
// %0 "$return_value"                             // user: %12
// %1 "x"                                         // users: %7, %3
// %2 "self"                                      // users: %5, %4
bb0(%0 : $*NonTrivial, %1 : $Int32, %2 : $*NonTrivialArrayByRef):
  debug_value %1 : $Int32, let, name "x", argno 1 // id: %3
  debug_value %2 : $*NonTrivialArrayByRef, var, name "self", argno 2, implicit, expr op_deref // id: %4
  %5 = begin_access [modify] [static] %2 : $*NonTrivialArrayByRef // users: %8, %7
  // function_ref _ZN20NonTrivialArrayByRefixEi
  %6 = function_ref @_ZN20NonTrivialArrayByRefixEi : $@convention(cxx_method) (Int32, @inout NonTrivialArrayByRef) -> UnsafeMutablePointer<NonTrivial> // user: %7
  %7 = apply %6(%1, %5) : $@convention(cxx_method) (Int32, @inout NonTrivialArrayByRef) -> UnsafeMutablePointer<NonTrivial> // user: %9
  end_access %5 : $*NonTrivialArrayByRef          // id: %8
  %9 = struct_extract %7 : $UnsafeMutablePointer<NonTrivial>, #UnsafeMutablePointer._rawValue // user: %10
  %10 = pointer_to_address %9 : $Builtin.RawPointer to [strict] $*NonTrivial // user: %11
  %11 = begin_access [read] [unsafe] %10 : $*NonTrivial // users: %13, %12
  copy_addr %11 to [initialization] %0 : $*NonTrivial // id: %12
  end_access %11 : $*NonTrivial                   // id: %13
  %14 = tuple ()                                  // user: %15
  return %14 : $()                                // id: %15
} // end sil function '$sSo20NonTrivialArrayByRefVySo0aB0Vs5Int32Vcig'
```
**Replacement Read Accessor**
```sil
// NonTrivialArrayByRef.subscript.read
sil shared [transparent] @$sSo20NonTrivialArrayByRefVySo0aB0Vs5Int32Vcir : $@yield_once @convention(method) (Int32, @inout NonTrivialArrayByRef) -> @yields @in_guaranteed NonTrivial {
// %0 "x"                                         // users: %6, %2
// %1 "self"                                      // users: %4, %3
bb0(%0 : $Int32, %1 : $*NonTrivialArrayByRef):
  debug_value %0 : $Int32, let, name "x", argno 1 // id: %2
  debug_value %1 : $*NonTrivialArrayByRef, var, name "self", argno 2, implicit, expr op_deref // id: %3
  %4 = begin_access [modify] [static] %1 : $*NonTrivialArrayByRef // users: %7, %6
  // function_ref _ZN20NonTrivialArrayByRefixEi
  %5 = function_ref @_ZN20NonTrivialArrayByRefixEi : $@convention(cxx_method) (Int32, @inout NonTrivialArrayByRef) -> UnsafeMutablePointer<NonTrivial> // user: %6
  %6 = apply %5(%0, %4) : $@convention(cxx_method) (Int32, @inout NonTrivialArrayByRef) -> UnsafeMutablePointer<NonTrivial> // user: %8
  end_access %4 : $*NonTrivialArrayByRef          // id: %7
  %8 = struct_extract %6 : $UnsafeMutablePointer<NonTrivial>, #UnsafeMutablePointer._rawValue // user: %9
  %9 = pointer_to_address %8 : $Builtin.RawPointer to [strict] $*NonTrivial // user: %10
  %10 = begin_access [read] [unsafe] %9 : $*NonTrivial // users: %12, %15, %11
  yield %10 : $*NonTrivial, resume bb1, unwind bb2 // id: %11

bb1:                                              // Preds: bb0
  end_access %10 : $*NonTrivial                   // id: %12
  %13 = tuple ()                                  // user: %14
  return %13 : $()                                // id: %14

bb2:                                              // Preds: bb0
  end_access %10 : $*NonTrivial                   // id: %15
  unwind                                          // id: %16
} // end sil function '$sSo20NonTrivialArrayByRefVySo0aB0Vs5Int32Vcir'
```

When invoked, both are inlined, but we can still see in the getter (first version) we `alloc_stack` space for a copy, 
then `copy_addr` into the location. When the read accessor is envoked, we simply just read the field of the struct we care about instead of copying the entire thing.

**Old main**
```sil
// main
sil @main : $@convention(c) (Int32, UnsafeMutablePointer<Optional<UnsafeMutablePointer<Int8>>>) -> Int32 {
bb0(%0 : $Int32, %1 : $UnsafeMutablePointer<Optional<UnsafeMutablePointer<Int8>>>):
  // ...
  %10 = alloc_stack $NonTrivial                   // users: %18, %33, %32, %21
  %11 = begin_access [modify] [static] %9 : $*NonTrivialArrayByRef // users: %14, %13
  // function_ref _ZN20NonTrivialArrayByRefixEi
  %12 = function_ref @_ZN20NonTrivialArrayByRefixEi : $@convention(cxx_method) (Int32, @inout NonTrivialArrayByRef) -> UnsafeMutablePointer<NonTrivial> // user: %13
  %13 = apply %12(%8, %11) : $@convention(cxx_method) (Int32, @inout NonTrivialArrayByRef) -> UnsafeMutablePointer<NonTrivial> // user: %15
  end_access %11 : $*NonTrivialArrayByRef         // id: %14
  %15 = struct_extract %13 : $UnsafeMutablePointer<NonTrivial>, #UnsafeMutablePointer._rawValue // user: %16
  %16 = pointer_to_address %15 : $Builtin.RawPointer to [strict] $*NonTrivial // user: %17
  %17 = begin_access [read] [unsafe] %16 : $*NonTrivial // users: %19, %18
  copy_addr %17 to [initialization] %10 : $*NonTrivial // id: %18
  end_access %17 : $*NonTrivial                   // id: %19
  %20 = tuple ()
  %21 = struct_element_addr %10 : $*NonTrivial, #NonTrivial.f // user: %22
  %22 = load %21 : $*Int32                        // user: %25
  // ..
} // end sil function 'main'
```
**New main**
```sil
sil @main : $@convention(c) (Int32, UnsafeMutablePointer<Optional<UnsafeMutablePointer<Int8>>>) -> Int32 {
bb0(%0 : $Int32, %1 : $UnsafeMutablePointer<Optional<UnsafeMutablePointer<Int8>>>):
  // ..
  %5 = function_ref @_ZN20NonTrivialArrayByRefC1Ev : $@convention(c) () -> @out NonTrivialArrayByRef // user: %6
  %6 = apply %5(%3) : $@convention(c) () -> @out NonTrivialArrayByRef
  %7 = integer_literal $Builtin.Int32, 0          // user: %8
  %8 = struct $Int32 (%7 : $Builtin.Int32)        // user: %12
  %9 = begin_access [modify] [dynamic] %3 : $*NonTrivialArrayByRef // users: %10, %21
  %10 = begin_access [modify] [static] %9 : $*NonTrivialArrayByRef // users: %13, %12
  // function_ref _ZN20NonTrivialArrayByRefixEi
  %11 = function_ref @_ZN20NonTrivialArrayByRefixEi : $@convention(cxx_method) (Int32, @inout NonTrivialArrayByRef) -> UnsafeMutablePointer<NonTrivial> // user: %12
  %12 = apply %11(%8, %10) : $@convention(cxx_method) (Int32, @inout NonTrivialArrayByRef) -> UnsafeMutablePointer<NonTrivial> // user: %14
  end_access %10 : $*NonTrivialArrayByRef         // id: %13
  %14 = struct_extract %12 : $UnsafeMutablePointer<NonTrivial>, #UnsafeMutablePointer._rawValue // user: %15
  %15 = pointer_to_address %14 : $Builtin.RawPointer to [strict] $*NonTrivial // user: %16
  %16 = begin_access [read] [unsafe] %15 : $*NonTrivial // users: %17, %19
  %17 = struct_element_addr %16 : $*NonTrivial, #NonTrivial.f // user: %18
  %18 = load %17 : $*Int32                        // user: %23
  // ...
} // end sil function 'main'
```